File: haskell/free-foil/src/Language/LambdaPi/Foil.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE QuantifiedConstraints #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE BlockArguments #-}

module Language.LambdaPi.Foil where

import Data.Map (Map)
import qualified Data.Map as Map
import Language.LambdaPi.Simple.Abs
import Language.LambdaPi.Simple.Lex (tokens)
import Language.LambdaPi.Simple.Layout (resolveLayout)
import Language.LambdaPi.Simple.Par (pProgram)
import Language.LambdaPi.Simple.Print (printTree)
import Unsafe.Coerce
import Data.Kind (Type)
import System.Exit (exitFailure)

type Id = String
type RawName = Id
type RawScope = [Id]

data {- kind -} S
  = {- type -} VoidS
  -- | {- type -} Singleton
  -- | {- type -} List

data Scope (n :: S) = UnsafeScope RawScope
data Name (n :: S) = UnsafeName RawName
    deriving (Show)
data NameBinder (n :: S) (l :: S) =
  UnsafeNameBinder (Name l)
    deriving (Show)

ppName :: Name n -> String
ppName (UnsafeName name) = name

-- UnsafeName "z" :: Name ["x", "y", "z"]
-- UnsafeNameBinder (UnsafeName "z")
--   :: NameBinder ["x"] ["x", "y", "z"]

emptyScope :: Scope VoidS
emptyScope = UnsafeScope []

extendScope :: NameBinder n l -> Scope n -> Scope l
extendScope (UnsafeNameBinder (UnsafeName name)) (UnsafeScope scope) =
  UnsafeScope (name : scope)

rawFreshName :: RawScope -> RawName
rawFreshName scope = head {- DO NOT WRITE LIKE ME -}
  [ name
  | n <- [1..]
  , let name = "x" <> show n
  , name `notElem` scope
  ]

withFreshBinder
  :: Scope n
  -> (forall l. NameBinder n l -> r)
  -> r
withFreshBinder (UnsafeScope scope) cont =
  cont binder
  where
    binder = UnsafeNameBinder (UnsafeName (rawFreshName scope))

nameOf :: NameBinder n l -> Name l
nameOf (UnsafeNameBinder name) = name

rawMember :: RawName -> RawScope -> Bool
rawMember i s = elem i s

member :: Name l -> Scope n -> Bool
member (UnsafeName name) (UnsafeScope s) = rawMember name s

data Expr n where
  VarE :: Name n -> Expr n
  AppE :: Expr n -> Expr n -> Expr n
  LamE :: NameBinder n l -> Expr l -> Expr n

-- >>> putStrLn $ ppExpr identity
-- λx1. x1
-- >>> putStrLn $ ppExpr two
-- λx1. λx2. x1(x1(x2))
ppExpr :: Expr n -> String
ppExpr expr = case expr of
  VarE name -> ppName name
  AppE e1 e2 -> ppExpr e1 ++ "(" ++ ppExpr e2 ++ ")"
  LamE x body -> "λ" ++ ppName (nameOf x) ++ ". " ++ ppExpr body


-- Distinct constraints
class ExtEndo (n :: S)

class (ExtEndo n => ExtEndo l ) => Ext (n:: S) (l :: S)
instance ( ExtEndo n => ExtEndo l ) => Ext n l

class Distinct (n :: S)
instance Distinct VoidS

type DExt n l = (Distinct l, Ext n l)

-- Safer scopes with distinct constraints
data DistinctEvidence ( n :: S) where
  Distinct :: Distinct n => DistinctEvidence n

unsafeDistinct :: DistinctEvidence n
unsafeDistinct = unsafeCoerce (Distinct :: DistinctEvidence VoidS)

data ExtEvidence ( n:: S) ( l :: S) where
  Ext :: Ext n l => ExtEvidence n l

unsafeExt :: ExtEvidence n l
unsafeExt = unsafeCoerce (Ext :: ExtEvidence VoidS VoidS)

withFresh :: Distinct n => Scope n
  -> (forall l . DExt n l => NameBinder n l -> r ) -> r
withFresh scope cont = withFreshBinder scope (\binder ->
  unsafeAssertFresh binder cont)

unsafeAssertFresh :: forall n l n' l' r. NameBinder n l
  -> (DExt n' l' => NameBinder n' l' -> r) -> r
unsafeAssertFresh binder cont =
  case unsafeDistinct @l' of
    -- #FIXME: when using originally @l' and @n' gives an error about type variables not in scope
    Distinct -> case unsafeExt @n' @l' of
      Ext -> cont (unsafeCoerce binder)

withRefreshed :: Distinct o => Scope o -> Name i
  -> (forall o'. DExt o o' => NameBinder o o' -> r) -> r
withRefreshed scope name cont = if member name scope
  then withFresh scope cont
  else unsafeAssertFresh (UnsafeNameBinder name) cont

-- generic sinking
concreteSink :: DExt n l => Expr n -> Expr l
concreteSink = unsafeCoerce

class Sinkable (e :: S -> *) where
  sinkabilityProof :: (Name n -> Name l) -> e n -> e l

instance Sinkable Name where
  sinkabilityProof rename = rename

sink :: (Sinkable e, DExt n l) => e n -> e l
sink = unsafeCoerce

instance Sinkable Expr where
  sinkabilityProof rename (VarE v) = VarE (rename v)
  sinkabilityProof rename (AppE f e) = AppE (sinkabilityProof rename f) (sinkabilityProof rename e)
  sinkabilityProof rename (LamE binder body) = extendRenaming rename binder \rename' binder' ->
    LamE binder' (sinkabilityProof rename' body)

extendRenaming :: (Name n -> Name n') -> NameBinder n l
  -> (forall l'. (Name l -> Name l') -> NameBinder n' l' -> r ) -> r
extendRenaming _ (UnsafeNameBinder name) cont =
  cont unsafeCoerce (UnsafeNameBinder name)

-- CoSinkable (for patterns)

class CoSinkable (pattern :: S -> S -> Type) where
  coSinkabilityProof
    :: (Name n -> Name n')
    -> pattern n l
    -> (forall l'. (Name l -> Name l') -> pattern n' l' -> r)
    -> r

instance CoSinkable NameBinder where
  coSinkabilityProof _rename (UnsafeNameBinder name) cont = cont unsafeCoerce (UnsafeNameBinder name)

-- Substitution
data Substitution (e :: S -> *) (i :: S) (o :: S) =
  UnsafeSubstitution (forall n. Name n -> e n) (Map String (e o))

lookupSubst :: Substitution e i o -> Name i -> e o
lookupSubst (UnsafeSubstitution f env) (UnsafeName name) =
    case Map.lookup name env of
        Just ex -> ex
        Nothing -> f (UnsafeName name)

identitySubst :: (forall n. Name n -> e n) -> Substitution e i i
identitySubst f = UnsafeSubstitution f Map.empty

addSubst :: Substitution e i o -> NameBinder i i' -> e o -> Substitution e i' o
addSubst (UnsafeSubstitution f env) (UnsafeNameBinder (UnsafeName name)) ex = UnsafeSubstitution f (Map.insert name ex env)

addRename :: Substitution e i o -> NameBinder i i' -> Name o -> Substitution e i' o
addRename s@(UnsafeSubstitution f env) b@(UnsafeNameBinder (UnsafeName name1)) n@(UnsafeName name2)
    | name1 == name2 = UnsafeSubstitution f env
    | otherwise = addSubst s b (f n)

instance (Sinkable e) => Sinkable (Substitution e i) where
  sinkabilityProof rename (UnsafeSubstitution f env) =
    UnsafeSubstitution f (fmap (sinkabilityProof rename) env)


-- Substitute part
substitute :: Distinct o => Scope o -> Substitution Expr i o -> Expr i -> Expr o
substitute scope subst = \case
    VarE name -> lookupSubst subst name
    AppE f x -> AppE (substitute scope subst f) (substitute scope subst x)
    LamE binder body -> withRefreshed scope (nameOf binder) (\binder' ->
        let subst' = addRename (sink subst) binder (nameOf binder')
            scope' = extendScope binder' scope
            body' = substitute scope' subst' body in LamE binder' body'
        )

toFoilTerm :: (Distinct n) => Term -> Scope n -> Map String (Name n) -> Expr n
toFoilTerm (App a b) scope env = AppE (toFoilTerm a scope env) (toFoilTerm b scope env)
toFoilTerm (Lam (Ident str) body) scope env = withFresh scope (\s ->
  let scope' = extendScope s scope
      env' = Map.insert str (nameOf s) (fmap sink env) in LamE s (toFoilTerm body scope' env'))
toFoilTerm (Var (Ident str)) scope env = case Map.lookup str env of
  Just name -> VarE name
  Nothing -> error ("Unbound variable " ++ str)
-- #TODO: suppori Pi expressions
toFoilTerm (Pi _ _ _) scope env = error ("Not yet implemented")

whnf :: (Distinct n) => Scope n -> Expr n -> Expr n
whnf scope = \case
  AppE f arg ->
    case whnf scope f of
      LamE z body ->
        let subst = UnsafeSubstitution VarE (Map.insert (ppName (nameOf z)) arg Map.empty)
        in substitute scope subst body
      f' -> AppE f' arg
  t -> t

-- | Interpret a λΠ command.
interpretCommand :: Command -> IO ()
interpretCommand (CommandCompute term type_) =
      putStrLn ("  ↦ " ++ ppExpr (whnf emptyScope (toFoilTerm term emptyScope Map.empty)))
-- #TODO: add typeCheck
interpretCommand (CommandCheck term type_) =
      putStrLn ("Not yet implemented")

-- | Interpret a λΠ program.
interpretProgram :: Program -> IO ()
interpretProgram (AProgram typedTerms) = mapM_ interpretCommand typedTerms

-- | Default interpreter program.
-- Reads a λΠ program from the standard input and runs the commands.
foilMain :: IO ()
foilMain = do
  input <- getContents
  case pProgram (resolveLayout True (tokens input)) of
    Left err -> do
      putStrLn err
      exitFailure
    Right program -> interpretProgram program

File: scala/free-foil/src/main/scala/Foil.scala
import Foil.DExt
import scala.collection.immutable.IntMap

object Foil extends App {

  sealed trait Expr[N <: S]

  case class VarE[N <: S](name: Name[N])                                   extends Expr[N]
  case class AppE[N <: S](fun: Expr[N], arg: Expr[N])                      extends Expr[N]
  case class LamE[N <: S, L <: S](binder: NameBinder[N, L], body: Expr[L]) extends Expr[N]

  case class RawName(rawId: Int)

  case class RawScope(rawSet: Set[Int])

  val rawEmptyScope: RawScope = RawScope(Set.empty[Int])

  def rawFreshName(rawScope: RawScope): RawName = rawScope match {
    case RawScope(s) if s.isEmpty => RawName(0)
    case RawScope(s)              => RawName(s.max + 1)
  }

  def rawExtendScope(rawName: RawName, rawScope: RawScope): RawScope =
    RawScope(rawScope.rawSet + rawName.rawId)

  def rawMember(rawName: RawName)(rawScope: RawScope): Boolean =
    rawScope.rawSet.contains(rawName.rawId)

  case class RawSubst[A](rawMap: IntMap[A])

  def rawIdSubst[A] = IntMap.empty[A]

  def rawLookup[A](rawSubst: RawSubst[A])(rawName: RawName): Option[A] =
    rawSubst.rawMap.get(rawName.rawId)

  def rawExtendSubst[A](rawName: RawName, value: A, rawSubst: RawSubst[A]): RawSubst[A] =
    RawSubst(rawSubst.rawMap.updated(rawName.rawId, value))

  sealed trait S
  case class VoidS() extends S

  case class Name[N <: S](unsafeName: RawName)
  case class Scope[N <: S](unsafeScope: RawScope)

  val emptyScope = Scope[VoidS](rawEmptyScope)

  def member[L <: S, N <: S](name: Name[L])(scope: Scope[N]): Boolean =
    rawMember(name.unsafeName)(scope.unsafeScope)

  case class NameBinder[N <: S, L <: S](unsafeBinder: Name[L])

  def nameOf[N <: S, L <: S](nameBinder: NameBinder[N, L]): Name[L] =
    nameBinder.unsafeBinder

  def withFreshBinder[N <: S, R](scope: Scope[N])(cont: [L <: S] => NameBinder[N, L] => R): R =
    cont(NameBinder(Name(rawFreshName(scope.unsafeScope))))

  class Distinct[N <: S]
  class ExtEndo[N <: S]
  class Ext[N <: S, L <: S](implicit ev: ExtEndo[N] <:< ExtEndo[L])

  type DExt[N <: S, L <: S] = (Distinct[L], Ext[N, L])

  sealed trait DistinctEvidence[N <: S]
  case object Distinct extends DistinctEvidence[VoidS]
  val unsafeDistinct: DistinctEvidence[VoidS] = Distinct

  sealed trait ExtEvidence[N <: S, L <: S]
  case object Ext extends ExtEvidence[VoidS, VoidS]
  val unsafeExt: ExtEvidence[VoidS, VoidS] = Ext

  class Sinkable[E[_ <: S]] {
    def sinkabilityProof[N <: S, L <: S](rename: Name[N] => Name[L]): E[N] => E[L] = ???
  }

  object Sinkable {
    given nameSinkable: Sinkable[Name] = new Sinkable[Name]
    given exprSinkable: Sinkable[Expr] = new Sinkable[Expr]
    given substitutionSinkable[E[_ <: S], I <: S](
        using sinkable: Sinkable[E],
    ): Sinkable[[O <: S] =>> Substitution[E, I, O]] with {
      def sinkabilityProof[N <: S, L <: S](
          rename: Name[N] => Name[L],
          substitution: Substitution[E, I, N],
      ): Substitution[E, I, L] = ???
    }
  }

  def sink[E[_ <: S]: Sinkable, N <: S, L <: S](en: E[N])(using ev: DExt[N, L]): E[L] = en.asInstanceOf[E[L]]

  def unsafeAssertFresh[N <: S, L <: S, N1 <: S, L1 <: S, R](
      binder: NameBinder[N, L],
  )(cont: DExt[N1, L1] ?=> NameBinder[N1, L1] => R): R = {
    given DExt[N1, L1] = (unsafeDistinct.asInstanceOf[Distinct[L1]], unsafeExt.asInstanceOf[Ext[N1, L1]])
    cont(binder.asInstanceOf[NameBinder[N1, L1]])
  }

  def withFresh[N <: S: Distinct, R](scope: Scope[N])(cont: [L <: S] => DExt[N, L] ?=> NameBinder[N, L] => R): R = {
    withFreshBinder[N, R](scope)([L <: S] => (binder: NameBinder[N, L]) => unsafeAssertFresh(binder)(cont[L]))
  }

  def withRefreshed[O <: S, I <: S, R](scope: Scope[O], name: Name[I])(
      cont: [O1 <: S] => DExt[O, O1] ?=> NameBinder[O, O1] => R,
  ): R = {

    given Distinct[O] = unsafeDistinct.asInstanceOf[Distinct[O]]
    given DExt[O, O]  = (unsafeDistinct.asInstanceOf[Distinct[O]], unsafeExt.asInstanceOf[Ext[O, O]])
    if (member(name)(scope))
      withFresh(scope)(cont)
    else
      unsafeAssertFresh[O, I, O, I, R](NameBinder[O, I](name))(cont[I])
  }

  def extendScope[N <: S, L <: S](nameBinder: NameBinder[N, L], scope: Scope[N]): Scope[L] =
    Scope[L](rawExtendScope(nameBinder.unsafeBinder.unsafeName, scope.unsafeScope))

  case class Substitution[E[_ <: S], I <: S, O <: S](f: [N <: S] => Name[N] => E[N], env: IntMap[E[O]])

  def lookupSubst[E[_ <: S], I <: S, O <: S](subst: Substitution[E, I, O], name: Name[I]): E[O] =
    subst.env.get(name.unsafeName.rawId) match {
      case Some(e) => e
      case None    => subst.f(Name(RawName(name.unsafeName.rawId)))
    }

  def idSubst[E[_ <: S], I <: S](f: [N <: S] => Name[N] => E[N]): Substitution[E, I, I] =
    Substitution(f, IntMap.empty)

  def addSubst[E[_ <: S], I <: S, O <: S, I1 <: S](
      s: Substitution[E, I, O],
  )(b: NameBinder[I, I1])(e: E[O]): Substitution[E, I1, O] = s match {
    case Substitution(f, env) => Substitution(f, env)
  }

  def addRename[E[_ <: S], I <: S, I1 <: S, O <: S](
      s: Substitution[E, I, O],
      b: NameBinder[I, I1],
      n: Name[O],
  ): Substitution[E, I1, O] = (s, b, n) match {
    case (Substitution(f, env), NameBinder(Name(name1)), Name(name2)) =>
      if (name1 == name2)
        Substitution(f, env)
      else
        addSubst(s)(b)(f(n))
  }

  def substitute[O <: S, I <: S](scope: Scope[O], subst: Substitution[Expr, I, O], expr: Expr[I]): Expr[O] = expr match {
    case VarE(name) => lookupSubst(subst, name)
    case AppE(f, x) => AppE(substitute(scope, subst, f), substitute(scope, subst, x))
    case LamE(binder, body) => withRefreshed(scope, nameOf(binder))
      { [O1 <: S] => implicit dext: DExt[O, O1] => (binder_ : NameBinder[O, O1]) =>
      val subst_ = addRename(sink(subst), binder, nameOf(binder_))
      val scope_ = extendScope(binder_, scope)
      val body_ = substitute(scope_, subst_, body)
      LamE(binder_, body_)
    }
  }

}
